### PR TITLE
v10 Upgrade Handler

### PR DESCRIPTION
**Pull request by sampocs**
## Context and purpose of the change
Upgrade handler for v10 upgrade

## Brief Changelog
* Migrated tendermint consensus params [[source](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0-rc2/UPGRADING.md#xconsensus)]
* Migrated channel capability from stakeibc → icacontroller [[source](https://ibc.cosmos.network/main/migrations/v5-to-v6.html#upgrade-proposal)]
* Pruned expired tendermint clients [[source](https://ibc.cosmos.network/main/migrations/v6-to-v7.html#chains)]
* Migrated icacallback proto from google proto (deprecated) to cosmos proto
* Reduced STRD staking rewards. Double check me on these numbers!!
* Set MinInitialDepositRatio
* Removed an unnecessary setup call from past unit tests
* Cleaned up makefile and docs
* Added localhost client for ibc-go v7.1

### CODE REVIEW

File: app/upgrades/v10/upgrades.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	communityPoolSecurityBudgetProportion := sdk.MustNewDecFromStr(CommunityPoolSecurityBudgetProportion)
+
+	// Confirm proportions sum to 100
+	totalProportions := stakingProportion.
+		Add(communityPoolGrowthProportion).
+		Add(strategicReserveProportion).
+		Add(communityPoolSecurityBudgetProportion)
+
+	if !totalProportions.Equal(sdk.OneDec()) {
+		return fmt.Errorf("distribution proprotions do not sum to 1 (%v)", totalProportions)
```
Comment (riley-stride) [inline app/upgrades/v10/upgrades.go:113]: pico nit
```suggestion
		return fmt.Errorf("distribution proportions do not sum to 1 (%v)", totalProportions)
```

File: app/upgrades/v10/upgrades.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	ibctmmigrations "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint/migrations"
+
+	mintkeeper "github.com/Stride-Labs/stride/v9/x/mint/keeper"
+	minttypes "github.com/Stride-Labs/stride/v9/x/mint/types"
+	stakeibctypes "github.com/Stride-Labs/stride/v9/x/stakeibc/types"
+)
+
+var (
+	UpgradeName     = "v10"
+	EpochProvisions = sdk.NewDec(940_469_178)
```
Comment (riley-stride) [inline app/upgrades/v10/upgrades.go:30]: I think this should be a 13.8% reduction, not a 12.8% reduction:

```
1-(940,469,178/1,078,767,123) = 0.128
```

File: app/upgrades/v10/upgrades_test.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	suite.Run(t, new(UpgradeTestSuite))
+}
+
+func (s *UpgradeTestSuite) TestUpgrade() {
+	dummyUpgradeHeight := int64(5)
+
+	s.ConfirmUpgradeSucceededs("v10", dummyUpgradeHeight)
+
+	// Check mint parameters after upgrade
+	zeroPad := "00000000000000"
```
Comment (riley-stride) [inline app/upgrades/v10/upgrades_test.go:30]: Could you explain this?

File: app/upgrades/v10/upgrades.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+			cdc,
+			capabilityStoreKey,
+			capabilityKeeper,
+			stakeibctypes.ModuleName,
+		); err != nil {
+			return nil, errorsmod.Wrapf(err, "unable to migrate ICA channel capabilities")
+		}
+
+		ctx.Logger().Info("Pruning expired tendermint consensus states...")
+		if _, err := ibctmmigrations.PruneExpiredConsensusStates(ctx, cdc, clientKeeper); err != nil {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:72]: yup

File: app/upgrades/v10/upgrades_test.go (hunk 3 above)
Comment (sampocs) [reply app/upgrades/v10/upgrades_test.go:30]: When converting a decimal to a string, it tacks on the extra zeros, so I'm adding it to the test here for the comparison. The alternative is to use .Equals, but then it doesn't print the value of each so it's hard to debug

File: app/app.go [hunk 5]
```diff
@@ -1000,6 +1000,7 @@ func (app *StrideApp) RegisterAPIRoutes(apiSvr *api.Server, apiConfig config.API
 	tmservice.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
 
 	ModuleBasics.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
+	nodeservice.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
 }
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/app.go:1004]: I don't understand it super well, but it was mentioned [here](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0-rc2/UPGRADING.md#grpc)

File: app/upgrades/v10/upgrades.go (hunk 2 above)
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:30]: Good catch!! Updated to 929,681,506

File: app/upgrades/v10/upgrades.go (hunk 1 above)
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:113]: done

File: app/upgrades/v10/upgrades.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+			return nil, errorsmod.Wrapf(err, "unable to set MinInitialDepositRatio")
+		}
+
+		ctx.Logger().Info("v10 Upgrade Complete")
+
+		return vm, err
+	}
+}
+
+// Cut STRD staking rewards in half from 27.64% to 13.82%
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:95]: yeah I think it depends on whether we re-normalize, I'll update it after you and Aidan decide on the correct proportions

**Comment by riley-stride** [comment]
Should I re-review this @sampocs?

**Review by riley-stride — APPROVED** [review]

File: app/upgrades/v10/upgrades.go [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+	NewDistributorAddresses = map[string]string{
+		"stride":  "stride1w02dg74j8s38gqn6mvlr87hkvyv5rgp3cqe9se",
+		"gaia":    "stride1w0w0gr6u796y2mjl9fuqt66jqvk3j59jq3jtpg",
+		"osmosis": "stride1mfg5ck02tlyzdtdpaj70ngtgjs2vuawtkfz7xd",
+		"juno":    "stride1ral4dsqk0nzyqlwtuyxgavfvx8hegml7u0rzx3",
+		"stars":   "stride1rm9nxc5pw3k5r5s6lm85k73mfp734nhnxq570g",
+		"evmos":   "stride1ej4e7x2hanmy6vrzrjh06g6dnfq5kxm73dmgsw",
+	}
+
+	RateLimitDurationHours = uint64(24)
```
Comment (asalzmann) [inline app/upgrades/v10/upgrades.go:79]: This seems aggressive - do you know what Osmosis uses?

File: app/upgrades/v10/upgrades.go [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+		if err := EnableRateLimits(ctx, accountKeeper, channelKeeper, ratelimitKeeper, stakeibcKeeper); err != nil {
+			return nil, errorsmod.Wrapf(err, "unable to enable rate limits")
+		}
+
+		ctx.Logger().Info("v10 Upgrade Complete")
+		return vm, err
+	}
+}
+
+// Cut STRD staking rewards in half
```
[the remark this answers is not in the document]
Comment (asalzmann) [reply app/upgrades/v10/upgrades.go:177]: Makes it easier for external parties to review

**Review by asalzmann — APPROVED** [review]

File: app/upgrades/v10/upgrades.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:177]: done

File: app/upgrades/v10/upgrades.go [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+			cdc,
+			capabilityStoreKey,
+			capabilityKeeper,
+			stakeibctypes.ModuleName,
+		); err != nil {
+			return nil, errorsmod.Wrapf(err, "unable to migrate ICA channel capabilities")
+		}
+
+		ctx.Logger().Info("Adding localhost IBC client...")
+		AddLocalhostIBCClient(ctx, clientKeeper)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:132]: It determines the total escrow'd tokens. It's done automatically

File: app/upgrades/v10/upgrades.go [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+) upgradetypes.UpgradeHandler {
+	return func(ctx sdk.Context, _ upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
+		ctx.Logger().Info("Starting upgrade v10...")
+
+		ctx.Logger().Info("Migrating tendermint consensus params from x/params to x/consensus...")
+		legacyParamSubspace := paramsKeeper.Subspace(baseapp.Paramspace).WithKeyTable(paramstypes.ConsensusParamsKeyTable())
+		baseapp.MigrateParams(ctx, legacyParamSubspace, &consensusParamsKeeper)
+
+		ctx.Logger().Info("Migrating ICA channel capabilities...")
+		if err := icacontrollermigrations.MigrateICS27ChannelCapability(
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:121]: we don't differentiate between the different hops, it all happens under the hood. I can add a clarifying comment though

File: app/upgrades/v10/upgrades.go (hunk 7 above)
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:79]: I don't. I agree it's aggressive. Wdyt it should be?

File: app/upgrades/v10/upgrades.go [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+		if hostZone.FeeAccount == nil || hostZone.FeeAccount.Address == "" {
+			return stakeibctypes.ErrICAAccountNotFound
+		}
+
+		ratelimitKeeper.SetWhitelistedAddressPair(ctx, ratelimittypes.WhitelistedAddressPair{
+			Sender:   hostZone.Address,
+			Receiver: hostZone.DelegationAccount.Address,
+		})
+
+		rewardCollectorAddress := accountKeeper.GetModuleAccount(ctx, stakeibctypes.RewardCollectorName).GetAddress()
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:352]:
```bash
>>> stridedm q stakeibc module-address reward_collector
addr: stride178jw99dmgyaqkmn5meevmcak27qte0gnymrztv
```

Felt safer imo to use the same code that we use in-protocol to send funds there than to hard code the address

File: app/upgrades/v10/upgrades.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:132]: escrow amount was also verified by localstride

File: app/upgrades/v10/upgrades.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (asalzmann) [reply app/upgrades/v10/upgrades.go:359]: Sounds good, we should add this to the localstride checklist (query the RL and make sure the address lines up)

File: app/upgrades/v10/upgrades.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:128]: imo this is better suited as a docstring comment than an alias of an import. added in latest commit

[the remark this answers is not in the document]
Comment (asalzmann) [reply app/upgrades/v10/upgrades.go:128]: What does alias of import mean? 

I'm suggesting breaking the logic up into functions, which we do below for other migrations

[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:128]: `MigrateICS27ChannelCapability` isn't our own custom function - it's a function imported from the icacontroller migrations package. I thought you were suggesting importing the `icacontrollermigrations` type differently

[the remark this answers is not in the document]
Comment (asalzmann) [reply app/upgrades/v10/upgrades.go:128]: Ah sorry, my comments was unclear. I was suggesting batching the logic associated with each migration (v5->v6, v6->v7, v7->v7.1) into functions, like so

```
ctx.Logger().Info("Running ibc v5->v6 migration...")
if err := IBCv5Tov6Migration(ctx, ...); err != nil {
	return nil, errorsmod.Wrapf(err, "unable to migrate ibc-go from v5 to v6")
}
```
and putting the relevant logic in `IBCv5Tov6Migration` 

A comment around which migration it is would also be fine. Anyway, it's a just a small thing

File: app/upgrades/v10/upgrades.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go]: done

File: app/upgrades/v10/upgrades.go [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+	stakeibckeeper "github.com/Stride-Labs/stride/v9/x/stakeibc/keeper"
+	stakeibctypes "github.com/Stride-Labs/stride/v9/x/stakeibc/types"
+
+	cosmosproto "github.com/cosmos/gogoproto/proto"
+	deprecatedproto "github.com/golang/protobuf/proto" //nolint:staticcheck
+)
+
+var (
+	UpgradeName     = "v10"
+	EpochProvisions = sdk.NewDec(929_681_506)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v10/upgrades.go:47]: updated [here](https://github.com/Stride-Labs/stride/pull/805/commits/37b7282eb012bb80129a708f38bd15e15d2993a4) per vishal's suggestion

**Pull request merged**